File: backend/src/models/attack.ts
import { Target, CreateTargetRequest } from "./target";

export enum AttackFrequency {
    LOW = "low",
    MEDIUM = "medium",
    HIGH = "high",
    VERY_HIGH = "very_high",
}

export enum AttackDanger {
    LOW = "low",
    MEDIUM = "medium",
    HIGH = "high",
    CRITICAL = "critical",
}

export enum AttackType {
    VOLUMETRIC = "volumetric",
    PROTOCOL = "protocol",
    APPLICATION = "application",
    AMPLIFICATION = "amplification",
}

export enum Protocol {
    HTTP = "http",
    HTTPS = "https",
    TCP = "tcp",
    UDP = "udp",
    SSH = "ssh",
    DNS = "dns",
}

export interface Attack {
    id: string;
    name: string;
    frequency: AttackFrequency;
    danger: AttackDanger;
    attack_type: AttackType;
    source_ips: string[]; // Массив IP адресов источников
    affected_ports: number[]; // Массив затронутых портов
    mitigation_strategies: string[]; // Массив стратегий защиты
    created_at: Date;
    updated_at: Date;
}

export interface CreateAttackRequest {
    name: string;
    frequency: AttackFrequency;
    danger: AttackDanger;
    attack_type: AttackType;
    source_ips?: string[];
    affected_ports?: number[];
    mitigation_strategies?: string[];
    targets?: CreateTargetRequest[];
}

export interface UpdateAttackRequest {
    name?: string;
    frequency?: AttackFrequency;
    danger?: AttackDanger;
    attack_type?: AttackType;
    source_ips?: string[];
    affected_ports?: number[];
    mitigation_strategies?: string[];
}

export interface AttackWithTargets extends Attack {
    targets: Target[];
}

File: frontend/models/attack.py
from dataclasses import dataclass, field
from typing import List, Dict, Any
from datetime import datetime
from utils.helpers import generate_id, get_current_timestamp


@dataclass
class Target:
    target_ip: str = ""
    target_domain: str = ""
    port: int = 80
    protocol: str = "tcp"
    tags: List[str] = field(default_factory=list)


@dataclass
class Attack:
    name: str
    frequency: str = "high"
    danger: str = "high"
    attack_type: str = "amplification"
    source_ips: List[str] = field(default_factory=list)
    affected_ports: List[int] = field(default_factory=list)
    mitigation_strategies: List[str] = field(default_factory=list)
    targets: List[Target] = field(default_factory=list)
    id: str = field(default_factory=generate_id)
    created_at: str = field(default_factory=get_current_timestamp)
    updated_at: str = field(default_factory=get_current_timestamp)

    def to_dict(self) -> Dict[str, Any]:
        """Конвертация в словарь для сохранения"""
        return {
            "id": self.id,
            "name": self.name,
            "frequency": self.frequency,
            "danger": self.danger,
            "attack_type": self.attack_type,
            "source_ips": self.source_ips,
            "affected_ports": self.affected_ports,
            "mitigation_strategies": self.mitigation_strategies,
            "targets": [target.__dict__ for target in self.targets],
            "created_at": self.created_at,
            "updated_at": self.updated_at
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Attack':
        """Создание объекта из словаря"""
        attack = cls(
            name=data["name"],
            frequency=data["frequency"],
            danger=data["danger"],
            attack_type=data["attack_type"],
            source_ips=data["source_ips"],
            affected_ports=data["affected_ports"],
            mitigation_strategies=data["mitigation_strategies"],
            id=data["id"],
            created_at=data["created_at"],
            updated_at=data["updated_at"]
        )

        # Конвертация targets
        attack.targets = [
            Target(
                target_ip=target_data.get("target_ip", ""),
                target_domain=target_data.get("target_domain", ""),
                port=target_data.get("port", 80),
                protocol=target_data.get("protocol", "tcp"),
                tags=target_data.get("tags", [])
            ) for target_data in data.get("targets", [])
        ]

        return attack